File: backend/src/routes/researchPlan.js
/**
 * Research Plan generation route
 * POST /api/research-plan
 */

import express from 'express'
import {
  buildAcademicResearchPlanMessages,
  generateAcademicResearchPlan,
} from '../services/academicResearchPlanService.js'
import { buildResearchPlanMessages, generateResearchPlan } from '../services/researchPlanService.js'
import { streamChat } from '../services/streamChatService.js'
import { createSseStream, getSseConfig } from '../utils/sse.js'

const router = express.Router()

/**
 * POST /api/research-plan
 * Generate a structured deep research plan
 *
 * Request body:
 * {
 *   "provider": "gemini" | "openai" | "openai_compatibility" | "siliconflow" | "glm" | "modelscope" | "kimi",
 *   "message": "User message about research",
 *   "apiKey": "API key for the provider",
 *   "baseUrl": "Custom base URL (optional)",
 *   "model": "model-name" (optional)
 * }
 *
 * Response:
 * {
 *   "plan": "JSON string of the research plan"
 * }
 */
router.post('/research-plan', async (req, res) => {
  try {
    const { provider, message, apiKey, baseUrl, model, researchType = 'general' } = req.body

    if (!provider) {
      return res.status(400).json({ error: 'Missing required field: provider' })
    }
    if (!message) {
      return res.status(400).json({ error: 'Missing required field: message' })
    }
    if (!apiKey) {
      return res.status(400).json({ error: 'Missing required field: apiKey' })
    }

    const supportedProviders = [
      'gemini',
      'openai',
      'openai_compatibility',
      'siliconflow',
      'glm',
      'modelscope',
      'kimi',
      'nvidia',
    ]
    if (!supportedProviders.includes(provider)) {
      return res.status(400).json({
        error: `Unsupported provider: ${provider}. Supported: ${supportedProviders.join(', ')}`,
      })
    }

    console.log(`[API] generateResearchPlan: provider=${provider}, researchType=${researchType}`)

    // Use academic research plan service for academic research
    const planGenerator =
      researchType === 'academic' ? generateAcademicResearchPlan : generateResearchPlan

    console.log(
      `[API] Selected plan generator: ${researchType === 'academic' ? 'Academic' : 'General'}`,
    )

    const plan = await planGenerator(provider, message, apiKey, baseUrl, model)

    res.json({ plan })
  } catch (error) {
    console.error('[API] Research plan generation error:', error)
    res.status(500).json({
      error: 'Failed to generate research plan',
      message: error.message,
    })
  }
})

/**
 * POST /api/research-plan-stream
 * Stream a structured deep research plan via SSE
 */
router.post('/research-plan-stream', async (req, res) => {
  try {
    const {
      provider,
      message,
      apiKey,
      baseUrl,
      model,
      responseFormat,
      thinking,
      temperature,
      top_k,
      top_p,
      frequency_penalty,
      presence_penalty,
      contextMessageLimit,
      researchType = 'general',
    } = req.body

    if (!provider || !message) {
      return res.status(400).json({ error: 'Missing required fields: provider, message' })
    }
    if (!apiKey) {
      return res.status(400).json({ error: 'Missing required field: apiKey' })
    }

    const supportedProviders = [
      'gemini',
      'openai',
      'openai_compatibility',
      'siliconflow',
      'glm',
      'modelscope',
      'kimi',
      'nvidia',
    ]
    if (!supportedProviders.includes(provider)) {
      return res.status(400).json({
        error: `Unsupported provider: ${provider}. Supported: ${supportedProviders.join(', ')}`,
      })
    }

    console.log(`[API] researchPlanStream: provider=${provider}, researchType=${researchType}`)

    const sse = createSseStream(res, getSseConfig())
    sse.writeComment('ok')

    const controller = new AbortController()
    req.on('aborted', () => controller.abort())
    res.on('close', () => {
      if (!res.writableEnded && !res.writableFinished) {
        controller.abort()
      }
    })

    const resolvedResponseFormat =
      responseFormat ?? (provider !== 'gemini' ? { type: 'json_object' } : undefined)
    const resolvedThinking =
      thinking ??
      (provider === 'glm' || provider === 'modelscope' ? { type: 'disabled' } : undefined)

    // Select appropriate prompt builder based on research type
    // Import for buildAcademicResearchPlanMessages is needed if not already present
    // Since we can't easily add global imports here without potentially breaking things or duplicates,
    // we should rely on existing imports. Currently generateAcademicResearchPlan handles the whole flow non-streaming.
    // However, research-plan-stream uses streamChat service directly.
    // We need to conditionally use the academic prompt builder.

    // NOTE: We need to import buildAcademicResearchPlanMessages at the top of the file first.
    // For now, let's assume the import will be added in a separate step or we add it here if possible.
    // Actually, looking at the imports: import { generateAcademicResearchPlan } from '../services/academicResearchPlanService.js'
    // It doesn't export buildAcademicResearchPlanMessages. I need to update the import first.

    const isAcademic = researchType === 'academic'
    const promptBuilder = isAcademic ? buildAcademicResearchPlanMessages : buildResearchPlanMessages
    const promptMessages = promptBuilder(message)

    console.log(`[API] Streaming research plan with type: ${isAcademic ? 'Academic' : 'General'}`)
    for await (const chunk of streamChat({
      provider,
      apiKey,
      baseUrl,
      model,
      messages: promptMessages,
      responseFormat: resolvedResponseFormat,
      thinking: resolvedThinking,
      temperature,
      top_k,
      top_p,
      frequency_penalty,
      presence_penalty,
      contextMessageLimit,
      signal: controller.signal,
    })) {
      sse.sendEvent(chunk)
    }

    sse.close()
  } catch (error) {
    console.error('[API] researchPlanStream error:', error)
    if (!res.headersSent) {
      res.status(500).json({
        error: 'Failed to stream research plan',
        message: error.message,
      })
    } else {
      res.write(`data: ${JSON.stringify({ type: 'error', error: error.message })}\n\n`)
      res.end()
    }
  }
})

export default router

File: backend-python/src/routes/research_plan.py
"""
Research plan API routes.
"""

from __future__ import annotations

import json
from typing import AsyncGenerator
import asyncio

from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
from sse_starlette.sse import EventSourceResponse

from ..providers import is_provider_supported
from ..services.research_plan import (
    generate_academic_research_plan,
    generate_research_plan,
    stream_generate_academic_research_plan,
    stream_generate_research_plan,
)


router = APIRouter(tags=["research-plan"])


@router.post("/research-plan")
async def research_plan(request: Request) -> JSONResponse:
    body = await request.json()
    provider = body.get("provider")
    message = body.get("message")
    api_key = body.get("apiKey")
    base_url = body.get("baseUrl")
    model = body.get("model")
    research_type = body.get("researchType") or "general"

    if not provider:
        return JSONResponse(status_code=400, content={"error": "Missing required field: provider"})
    if not message:
        return JSONResponse(status_code=400, content={"error": "Missing required field: message"})
    if not api_key:
        return JSONResponse(status_code=400, content={"error": "Missing required field: apiKey"})
    if not is_provider_supported(provider):
        return JSONResponse(status_code=400, content={"error": f"Unsupported provider: {provider}"})

    if research_type == "academic":
        plan = await generate_academic_research_plan(
            provider=provider,
            user_message=message,
            api_key=api_key,
            base_url=base_url,
            model=model,
        )
    else:
        plan = await generate_research_plan(
            provider=provider,
            user_message=message,
            api_key=api_key,
            base_url=base_url,
            model=model,
        )
    return JSONResponse(content={"plan": plan})


@router.post("/research-plan-stream")
async def research_plan_stream(request: Request) -> EventSourceResponse:
    body = await request.json()
    provider = body.get("provider")
    message = body.get("message")
    api_key = body.get("apiKey")
    base_url = body.get("baseUrl")
    model = body.get("model")
    thinking = body.get("thinking")
    temperature = body.get("temperature")
    top_k = body.get("top_k")
    top_p = body.get("top_p")
    frequency_penalty = body.get("frequency_penalty")
    presence_penalty = body.get("presence_penalty")
    research_type = body.get("researchType") or "general"

    if not provider or not message:
        return EventSourceResponse(
            _error_stream("Missing required fields: provider, message"),
            media_type="text/event-stream",
        )
    if not api_key:
        return EventSourceResponse(
            _error_stream("Missing required field: apiKey"),
            media_type="text/event-stream",
        )
    if not is_provider_supported(provider):
        return EventSourceResponse(
            _error_stream(f"Unsupported provider: {provider}"),
            media_type="text/event-stream",
        )

    async def event_generator() -> AsyncGenerator[dict[str, str], None]:
        try:
            # Choose the appropriate streaming function based on research type
            if research_type == "academic":
                stream_func = stream_generate_academic_research_plan(
                    provider=provider,
                    user_message=message,
                    api_key=api_key,
                    base_url=base_url,
                    model=model,
                    temperature=temperature,
                    top_p=top_p,
                    top_k=top_k,
                    frequency_penalty=frequency_penalty,
                    presence_penalty=presence_penalty,
                    thinking=thinking,
                )
            else:
                stream_func = stream_generate_research_plan(
                    provider=provider,
                    user_message=message,
                    api_key=api_key,
                    base_url=base_url,
                    model=model,
                    temperature=temperature,
                    top_p=top_p,
                    top_k=top_k,
                    frequency_penalty=frequency_penalty,
                    presence_penalty=presence_penalty,
                    thinking=thinking,
                )

            async for event in stream_func:
                if await request.is_disconnected():
                    break
                yield {"data": json.dumps(event, ensure_ascii=False)}
        except asyncio.CancelledError:
            return

    return EventSourceResponse(event_generator(), media_type="text/event-stream")


async def _error_stream(message: str) -> AsyncGenerator[dict[str, str], None]:
    yield {"data": json.dumps({"type": "error", "error": message}, ensure_ascii=False)}
